tmt/steps/report/junit.py

```diff
@@ -58,40 +67,230 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
 
 
+class ResultWrapper:
+    """
+    Adds possibility to wrap the ``tmt.Result`` and dynamically add more attributes which get
+    available inside the template context.
+    """
+
+    class PropertyDict(TypedDict):
+        """ Defines a property dict, which gets propagated into the final template properties. """
+
+        name: str
+        value: str
+
+    def __init__(self, wrapped: tmt.Result) -> None:
+        self._wrapped = wrapped
+        self._properties: dict[str, str] = {}
+
+    @property
+    def properties(self) -> list[PropertyDict]:
+        return [{'name': k, 'value': v} for k, v in self._properties.items()]
+
+    def set_properties(self, keyval: dict[str, str]) -> None:
+        self._properties = keyval
+
+    def __getattr__(self, name: str) -> Any:
+        """ Returns an attribute of a wrapped ``tmt.Result`` instance """
+        return getattr(self._wrapped, name)
+
+
+def _add_result_properties(result_instance: tmt.Result) -> ResultWrapper:
+    """ Dynamically decorate the ``tmt.Result`` instance - add ``properties`` attribute """
+    return ResultWrapper(result_instance)
+
+
+class ResultsContext:
+    """
+    A class which keeps the results context (especially the result summary) for
+    JUnit template. It wraps all the ``tmt.Result`` instances into the ``ResultWrapper``.
+    """
+
+    def __init__(self, results: list[tmt.Result]) -> None:
+        # Decorate all the tmt.Results with more attributes
+        self._results: list[ResultWrapper] = [_add_result_properties(r) for r in results]
+
+    def __iter__(self) -> Iterator[ResultWrapper]:
+        """ Possibility to iterate over results by iterating an instance """
+        return iter(self._results)
+
+    def __len__(self) -> int:
+        """ Returns the number of results """
+        return len(self._results)
+
+    @functools.cached_property
+    def passed(self) -> list[ResultWrapper]:
+        """ Returns results of passed tests """
+        return [r for r in self._results if r.result == ResultOutcome.PASS]
+
+    @functools.cached_property
+    def skipped(self) -> list[ResultWrapper]:
+        """ Returns results of skipped tests """
+        return [r for r in self._results if r.result == ResultOutcome.INFO]
+
+    @functools.cached_property
+    def failed(self) -> list[ResultWrapper]:
+        """ Returns results of failed tests """
+        return [r for r in self._results if r.result == ResultOutcome.FAIL]
+
+    @functools.cached_property
+    def errored(self) -> list[ResultWrapper]:
+        """ Returns results of tests with error/warn outcome """
+        return [r for r in self._results if r.result in [
+                ResultOutcome.ERROR,
+                ResultOutcome.WARN]]
+
+    @functools.cached_property
+    def duration(self) -> int:
+        """ Returns the total duration of all tests in seconds """
+        durations: list[int] = [duration_to_seconds(r.duration) or 0 for r in self._results]
+        return sum(durations)
```

Comment:
`durations` does not seem to be needed as a standalone variable, and as a list, because `sum()` can take an iterable, including a generator: `sum(duration_to_seconds(r.duration) or 0 for r in self._results)`. The list of durations does not need to materialize, ever.

Reply:
Hm, maybe `self._results` is mistyped, who knows. Let' see what comes out of it, I believe we will be able to sort it out.

Reply:
I've pushed the change. Could you please take a look and help me resolve the error?

Reply:
```diff
diff --git a/tmt/steps/report/junit.py b/tmt/steps/report/junit.py
index db09ab40..f6fe4a1c 100644
--- a/tmt/steps/report/junit.py
+++ b/tmt/steps/report/junit.py
@@ -1,7 +1,7 @@
 import dataclasses
 import functools
 from collections.abc import Iterator
-from typing import TYPE_CHECKING, Any, Optional, TypedDict, overload
+from typing import TYPE_CHECKING, Any, Optional, TypedDict, cast, overload
 
 from jinja2 import FileSystemLoader, select_autoescape
 
@@ -145,7 +145,11 @@ class ResultsContext:
     @functools.cached_property
     def duration(self) -> int:
         """ Returns the total duration of all tests in seconds """
-        return sum(duration_to_seconds(r.duration) or 0 for r in self._results)
+        # cast: mypy does not understand the proxy-ness of `ResultWrapper`. `r.duration`
+        # will exists, therefore adding a `cast` to convince mypy the list is pretty much
+        # nothing but the list of results.
+        return sum(
+            duration_to_seconds(r.duration) or 0 for r in cast(list[tmt.Result], self._results))
 
 
 def make_junit_xml(
```